src/world/Units/Unit.cpp

```diff
@@ -1063,6 +1063,81 @@ bool Unit::hasAurasWithId(uint32_t* auraId)
     return false;
 }
 
+bool Unit::hasAuraWithAuraEffect(AuraEffect type) const
+{
+    for (uint32_t i = MAX_TOTAL_AURAS_START; i < MAX_TOTAL_AURAS_END; ++i)
+    {
+        if (m_auras[i] == nullptr)
+            continue;
+        if (m_auras[i]->GetSpellInfo()->HasEffectApplyAuraName(type))
+            return true;
+    }
+    return false;
+}
+
+bool Unit::hasAuraState(AuraState state, SpellInfo* spellInfo, Unit* caster) const
+{
+    if (caster != nullptr && spellInfo != nullptr && caster->hasAuraWithAuraEffect(SPELL_AURA_IGNORE_TARGET_AURA_STATE))
+    {
+        for (uint32_t i = MAX_TOTAL_AURAS_START; i < MAX_TOTAL_AURAS_END; ++i)
+        {
+            if (caster->m_auras[i] == nullptr)
+                continue;
+            if (!caster->m_auras[i]->GetSpellInfo()->HasEffectApplyAuraName(SPELL_AURA_IGNORE_TARGET_AURA_STATE))
+                continue;
+            if (caster->m_auras[i]->GetSpellInfo()->isAffectingSpell(spellInfo))
+                return true;
+        }
+    }
+    return unitData()->aura_state & (1 << (state - 1));
+}
+
+void Unit::addAuraState(AuraState state)
+{
+    if (!(unitData()->aura_state & (1 << (state - 1))))
+    {
+        write(unitData()->aura_state, unitData()->aura_state | uint32_t(1 << (state - 1)));
+        if (IsPlayer())
+        {
+            // Activate passive spells which require this aurastate
+            SpellSet const playerSpellMap = static_cast<Player*>(this)->mSpells;
+            for (const uint32_t spellId : playerSpellMap)
+            {
+                // Skip deleted spells, i.e. spells with lower rank than the current rank
+                SpellSet::const_iterator deletedSpell = static_cast<Player*>(this)->mDeletedSpells.find(spellId);
+                if ((deletedSpell != static_cast<Player*>(this)->mDeletedSpells.end()))
+                    continue;
+                SpellInfo* spellInfo = sSpellCustomizations.GetSpellInfo(spellId);
+                if (spellInfo == nullptr || !spellInfo->IsPassive())
+                    continue;
+                if (spellInfo->getCasterAuraState() == uint32_t(state))
+                    CastSpell(this, spellId, true);
+            }
+        }
+    }
+}
+
+void Unit::removeAuraState(AuraState state)
+{
+    if (unitData()->aura_state & (1 << (state - 1)))
+    {
+        write(unitData()->aura_state, unitData()->aura_state & ~(1 << (state - 1)));
```

Comment:
Please isolate the write call into its own function and then call that, instead of mixing it with other logic.

Reply:
Same as mentioned a few lines above.
This function does not only write the unit data, it also removes locale saved Unit auras.
I would suggest to rename this function to removeAuraStateAndAuras and isolate the write(unitData()) part to removeAuraState function.